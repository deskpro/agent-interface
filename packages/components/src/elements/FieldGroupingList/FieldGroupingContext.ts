import { createContext } from "react";

export type FieldGroupingContextType = {
  selected: React.Key[];
  onToggleItem: (itemName: React.Key, e: React.SyntheticEvent) => void;
};
export const FieldGroupingContext = createContext<FieldGroupingContextType>({
  selected: [],
  onToggleItem: () => {}
});
