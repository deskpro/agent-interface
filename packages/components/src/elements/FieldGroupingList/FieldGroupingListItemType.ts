export interface FieldGroupingListItemType {
  id: React.Key;
  title: string;
  type: string;
  count: number;
  children?: FieldGroupingListItemType[];
}
