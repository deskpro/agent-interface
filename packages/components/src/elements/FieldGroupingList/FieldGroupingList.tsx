import * as React from "react";
import * as intersection from "lodash/intersection";

import "@deskpro/agent-interface-style/dist/components/dp-GroupedFields.css";
import ListItem from "./FieldGroupingListItem";
import getChildrenIds from "./childrenIds";
import { FieldGroupingContext } from "./FieldGroupingContext";
import { FieldGroupingListItemType } from "./FieldGroupingListItemType";

export type FieldGroupingListProps = {
  items: FieldGroupingListItemType[];
  onSelectChange: (itemId: React.Key, e: React.SyntheticEvent) => void;
};

const FieldGroupingList: React.FC<FieldGroupingListProps> = ({
  items,
  onSelectChange
}) => {
  const [selected, updatedSelected] = React.useState<React.Key[]>([]);
  const handleSelection = React.useCallback(
    (itemName: React.Key, e: React.SyntheticEvent) => {
      updatedSelected(
        selected.includes(itemName)
          ? selected.filter(i => i !== itemName)
          : selected.concat(itemName)
      );
      onSelectChange(itemName, e);
    },
    [onSelectChange, selected]
  );

  const contextValue = React.useMemo(
    () => ({ selected, onToggleItem: handleSelection }),
    [selected, handleSelection]
  );

  return (
    <FieldGroupingContext.Provider value={contextValue}>
      <div className="dp-Group-items">
        {items.map(({ children = [], ...item }) => {
          const childrenIds = getChildrenIds(children);
          const selectedChildren = intersection(childrenIds, selected);
          return (
            <ListItem
              {...item}
              isSelected={
                selected.includes(item.id) ||
                (children.length > 0 && selectedChildren.length > 0)
              }
              totalNodes={childrenIds.length}
              selectedNodes={selectedChildren.length}
            >
              {children}
            </ListItem>
          );
        })}
      </div>
    </FieldGroupingContext.Provider>
  );
};

export default FieldGroupingList;
