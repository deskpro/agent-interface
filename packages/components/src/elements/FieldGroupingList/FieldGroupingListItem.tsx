import * as React from "react";
import classNames from "classnames";
import { Manager, Reference, Popper } from "react-popper";

import { FieldGroupingListItemType } from "./FieldGroupingListItemType";
import { FieldGroupingContext } from "./FieldGroupingContext";
import SubItem from "./FieldGroupingSubItem";
import Badge from "../Badges/Badge";
import useMenu from "../../hooks/useMenu";
import useOutsideClick from "../../hooks/useOutsideClick";
import Arrow from "../../Button/Arrow";

export type FieldGroupingListItemProps = {
  id: React.Key;
  count: number;
  title: string;
  totalNodes?: number;
  selectedNodes?: number;
  isSelected?: boolean;
  children: FieldGroupingListItemType[];
};

const FieldGroupingListItem: React.FC<FieldGroupingListItemProps> = ({
  id,
  count,
  title,
  totalNodes,
  selectedNodes,
  isSelected = false,
  children
}) => {
  const { isVisible: menuIsVisible, toggleMenu, hideMenu } = useMenu();
  const badgeRef = React.useRef<HTMLSpanElement>(null);
  useOutsideClick(badgeRef, hideMenu);

  const { onToggleItem } = React.useContext(FieldGroupingContext);

  const arrow = (
    <Reference>
      {({ ref }) => (
        <Arrow
          isActive={menuIsVisible}
          ref={ref}
          onClick={e => {
            e.stopPropagation();
            toggleMenu();
          }}
        />
      )}
    </Reference>
  );

  return (
    <Badge
      type="group-filter"
      label={title}
      value={count}
      className={classNames({
        "is-active": isSelected,
        "Badge--extended": children.length > 0
      })}
      ref={badgeRef}
      onClick={e => {
        e.stopPropagation();
        onToggleItem(id, e);
      }}
    >
      <Manager>
        {!!totalNodes && !!selectedNodes ? (
          <span className="dp-Num-extended">
            <span>
              <span className="dp-Num">{selectedNodes}</span>/{totalNodes}
            </span>
            {arrow}
          </span>
        ) : (
          !!children.length && arrow
        )}
        {!!children.length && menuIsVisible && (
          <Popper
            placement="bottom-start"
            modifiers={{ offset: { offset: "0,12" } }}
          >
            {({ ref, style }) => (
              <ul className="dp-Menu" style={style} ref={ref}>
                {children.map(child => (
                  <SubItem key={child.id} item={child} />
                ))}
              </ul>
            )}
          </Popper>
        )}
      </Manager>
    </Badge>
  );
};

export default FieldGroupingListItem;
