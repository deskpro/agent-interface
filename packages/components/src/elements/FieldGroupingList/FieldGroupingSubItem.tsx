import * as React from "react";
import classNames from "classnames";
import * as intersection from "lodash/intersection";

import { FieldGroupingListItemType } from "./FieldGroupingListItemType";
import { FieldGroupingContext } from "./FieldGroupingContext";
import getChildrenIds from "./childrenIds";
import Badge from "../Badges/Badge";
import Icon from "../Icon/Icon";
import Arrow from "../../Button/Arrow";

const SubItem: React.FC<{ item: FieldGroupingListItemType }> = ({ item }) => {
  const { selected, onToggleItem } = React.useContext(FieldGroupingContext);
  const selectedChildren = intersection(
    getChildrenIds(item.children || []),
    selected
  );
  const [isExpanded, toggleExpanded] = React.useState(
    selected.includes(item.id) || selectedChildren.length > 0
  );

  return (
    <li
      key={item.id}
      className={classNames({
        "dp-Group-list": !!item.children && item.children.length > 0
      })}
    >
      <span className="dp-Group-item">
        <Icon name="subfield" size={12} />
        <Badge
          type="group-filter"
          className={classNames({ "is-active": selected.includes(item.id) })}
          label={item.title}
          value={item.count}
          onClick={e => {
            e.stopPropagation();
            onToggleItem(item.id, e);
          }}
        >
          {!!item.children && item.children.length > 0 && (
            <Arrow
              isActive={isExpanded}
              onClick={e => {
                e.stopPropagation();
                toggleExpanded(!isExpanded);
              }}
            />
          )}
        </Badge>
      </span>
      {!!item.children && item.children.length > 0 && isExpanded && (
        <ul className="dp-Group-sublist">
          {item.children.map(i => (
            <SubItem key={i.id} item={i} />
          ))}
        </ul>
      )}
    </li>
  );
};

export default SubItem;
