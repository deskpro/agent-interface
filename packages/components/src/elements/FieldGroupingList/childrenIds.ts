import * as flatMapDeep from "lodash/flatMapDeep";

import { FieldGroupingListItemType } from "./FieldGroupingListItemType";

const childrenMapper = (item: FieldGroupingListItemType) =>
  [item.id].concat((item.children || []).map(childrenMapper));

export default (children: FieldGroupingListItemType[]) =>
  flatMapDeep(children, childrenMapper);
