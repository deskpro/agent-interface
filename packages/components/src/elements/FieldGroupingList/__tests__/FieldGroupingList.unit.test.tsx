import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import FieldGroupingList from "../FieldGroupingList";

const items = [
  {
    id: 100,
    title: "All",
    count: 900
  },
  {
    id: 110,
    title: "Regulated",
    count: 80
  },
  {
    id: 105,
    title: "IT",
    count: 5
  },
  {
    type: "department",
    id: 123,
    title: "Sales",
    count: 100,
    children: [
      {
        type: "department",
        id: 124,
        title: "Presales",
        count: 50
      },
      {
        type: "department",
        id: 125,
        title: "Non-Profit",
        count: 50,
        children: [
          {
            type: "department",
            id: 126,
            title: "Charity",
            count: 40
          },
          {
            type: "department",
            id: 127,
            title: "Religious",
            count: 10
          }
        ]
      }
    ]
  },
  {
    type: "department",
    id: 987,
    title: "Support",
    count: 500
  },
  {
    id: 888,
    title: "HR",
    count: 25
  }
];

describe("<FieldGroupingList />", () => {
  it("should match its snapshot", () => {
    const onSelectChange = jest.fn();
    const { container } = render(
      <FieldGroupingList items={items} onSelectChange={onSelectChange} />
    );
    expect(container).toMatchSnapshot();
  });
});
