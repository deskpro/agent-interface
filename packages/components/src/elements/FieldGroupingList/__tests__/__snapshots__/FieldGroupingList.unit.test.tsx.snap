// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldGroupingList /> should match its snapshot 1`] = `
<div>
  <div
    class="dp-Group-items"
  >
    <span
      class="dp-Badge Badge--group-filter"
      role="button"
      tabindex="-1"
    >
      All
      <span
        class="dp-Num"
      >
        900
      </span>
    </span>
    <span
      class="dp-Badge Badge--group-filter"
      role="button"
      tabindex="-1"
    >
      Regulated
      <span
        class="dp-Num"
      >
        80
      </span>
    </span>
    <span
      class="dp-Badge Badge--group-filter"
      role="button"
      tabindex="-1"
    >
      IT
      <span
        class="dp-Num"
      >
        5
      </span>
    </span>
    <span
      class="dp-Badge Badge--extended Badge--group-filter"
      role="button"
      tabindex="-1"
    >
      Sales
      <span
        class="dp-Num"
      >
        100
      </span>
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
    <span
      class="dp-Badge Badge--group-filter"
      role="button"
      tabindex="-1"
    >
      Support
      <span
        class="dp-Num"
      >
        500
      </span>
    </span>
    <span
      class="dp-Badge Badge--group-filter"
      role="button"
      tabindex="-1"
    >
      HR
      <span
        class="dp-Num"
      >
        25
      </span>
    </span>
  </div>
</div>
`;
